### PR TITLE
used get function when accessing dictionaries. updated error message …

### PR DESCRIPTION
**Pull request by m-kovalsky**
…handling.

### CODE REVIEW

**Review by eisber — CHANGES_REQUESTED** [review]

File: src/sempy_labs/_ai.py [hunk 1]
```diff
@@ -116,8 +114,8 @@ def generate_measure_descriptions(
     )
 
     # Update the model to use the new descriptions
-    tom_server = fabric.create_tom_server(readonly=False, workspace=workspace)
-    m = tom_server.Databases.GetByName(dataset).Model
+    #with connect_semantic_model(dataset=dataset, workspace=workspace, readonly=False) as tom:
```
[the remark this answers is not in the document]
Comment (m-kovalsky) [reply src/sempy_labs/_ai.py:117]: to keep things consistent and leverage the functions in the tom package

File: src/sempy_labs/_connections.py [hunk 2]
```diff
@@ -60,16 +60,16 @@ def create_connection_cloud(
...
-            "Connection Path": o["connectionDetails"]["path"],
-            "Privacy Level": o["privacyLevel"],
-            "Credential Type": o["credentialDetails"]["credentialType"],
-            "Single Sign On Type": o["credentialDetails"]["singleSignOnType"],
-            "Connection Encryption": o["credentialDetails"]["connectionEncryption"],
-            "Skip Test Connection": o["credentialDetails"]["skipTestConnection"],
+            "Connection Id": o.get("id"),
+            "Connection Name": o.get("name"),
+            "Connectivity Type": o.get("connectivityType"),
+            "Connection Type": o.get("connectionDetails").get("type"),
```
[the remark this answers is not in the document]
Comment (m-kovalsky) [reply src/sempy_labs/_connections.py:66]: thanks. fixed this

File: src/sempy_labs/_list_functions.py [hunk 3]
```diff
@@ -29,44 +30,44 @@ def get_object_level_security(dataset: str, workspace: Optional[str] = None):
         A pandas dataframe showing the object level security for the semantic model.
     """
 
-    if workspace is None:
-        workspace_id = fabric.get_workspace_id()
-        workspace = fabric.resolve_workspace_name(workspace_id)
-
-    tom_server = fabric.create_tom_server(readonly=True, workspace=workspace)
-    m = tom_server.Databases.GetByName(dataset).Model
+    from .tom import connect_semantic_model
```
[the remark this answers is not in the document]
Comment (m-kovalsky) [reply src/sempy_labs/_list_functions.py:33]: fixed all instances of this

File: src/sempy_labs/_list_functions.py [hunk 4]
```diff
@@ -29,44 +30,44 @@ def get_object_level_security(dataset: str, workspace: Optional[str] = None):
...
-                    )
-                else:
-                    objectType = "Column"
-                    for cp in tp.ColumnPermissions:
+    with connect_semantic_model(dataset=dataset, readonly=True, workspace=workspace) as tom:
+
+        for r in tom.model.Roles:
+            for tp in r.TablePermissions:
+                if len(tp.FilterExpression) == 0:
+                    columnCount = len(tp.ColumnPermissions)
```
[the remark this answers is not in the document]
Comment (m-kovalsky) [reply src/sempy_labs/_list_functions.py:45]: i updated it to this

```python
columnCount = 0
try:
    columnCount = len(tp.ColumnPermissions)
except:
    pass
```

File: src/sempy_labs/_list_functions.py [hunk 5]
```diff
@@ -107,32 +106,34 @@ def list_tables(dataset: str, workspace: Optional[str] = None):
...
-            sourceExpression = t.RefreshPolicy.SourceExpression
+        for t in tom.model.Tables:
+            tableType = "Table"
+            rPolicy = bool(t.RefreshPolicy)
+            sourceExpression = None
+            if str(t.CalculationGroup) != "None":
+                tableType = "Calculation Group"
+            else:
+                for p in t.Partitions:
+                    if str(p.SourceType) == "Calculated":
```
[the remark this answers is not in the document]
Comment (m-kovalsky) [reply src/sempy_labs/_list_functions.py:119]: done

File: src/sempy_labs/_list_functions.py [hunk 6]
```diff
@@ -154,12 +155,10 @@ def list_annotations(dataset: str, workspace: Optional[str] = None):
...
     """
 
-    if workspace is None:
-        workspace_id = fabric.get_workspace_id()
-        workspace = fabric.resolve_workspace_name(workspace_id)
+    from .tom import connect_semantic_model
 
-    tom_server = fabric.create_tom_server(readonly=True, workspace=workspace)
-    m = tom_server.Databases.GetByName(dataset).Model
+    if workspace is None:
```
[the remark this answers is not in the document]
Comment (m-kovalsky) [reply src/sempy_labs/_list_functions.py:160]: updated this

File: src/sempy_labs/_list_functions.py [hunk 7]
```diff
@@ -171,183 +170,185 @@ def list_annotations(dataset: str, workspace: Optional[str] = None):
...
+        mName = tom.model.Name
+        for a in tom.model.Annotations:
+            objectType = "Model"
+            aName = a.Name
+            aValue = a.Value
             new_data = {
-                "Object Name": tName,
-                "Parent Object Name": mName,
+                "Object Name": mName,
+                "Parent Object Name": "N/A",
```
[the remark this answers is not in the document]
Comment (m-kovalsky) [reply src/sempy_labs/_list_functions.py:182]: updated to None

File: src/sempy_labs/_list_functions.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (eisber) [reply src/sempy_labs/_list_functions.py:45]: catch exception w/o type and instead of checking for None is usually not a good idea (e.g. what if there's an actual exception for some other reason, it will be very hard to find). let me rephrase the question: have you encounter this property to be None (or is it documented)? if it's not documented and you haven't encountered I think it's fine to just keep the len(...) w/o it. If it can be None, better to check for none using tp.ColumnPermission is None:
